wfe/wfe.go

```diff
@@ -1775,6 +1848,114 @@ func (wfe *WebFrontEndImpl) orderForDisplay(
 	return result
 }
 
+// RenewalInfo implements ACME Renewal Info (ARI)
+func (wfe *WebFrontEndImpl) RenewalInfo(_ context.Context, response http.ResponseWriter, request *http.Request) {
+	if request.Method == http.MethodPost {
+		//wfe.UpdateRenewal(context.TODO(), response, request)
+		wfe.sendError(acme.InternalErrorProblem("POSTing to RenewalInfo has not been implemented yet"), response)
+		return
+	}
+
+	if len(request.URL.Path) == 0 {
+		wfe.sendError(acme.NotFoundProblem("Must specify a request path"), response)
+		return
+	}
+
+	certID, err := wfe.parseCertID(request.URL.Path)
+	if err != nil {
+		wfe.sendError(acme.MalformedProblem(fmt.Sprintf("Parsing ARI CertID failed: %s", err)), response)
+		return
+	}
+
+	renewalInfo, err := wfe.determineARIWindow(context.TODO(), certID)
+	if err != nil {
+		wfe.sendError(acme.InternalErrorProblem(fmt.Sprintf("Error determining renewal window: %s", err)), response)
+		return
+	}
+
+	response.Header().Set("Retry-After", strconv.Itoa(int(6*time.Hour/time.Second)))
+	err = wfe.writeJSONResponse(response, http.StatusOK, renewalInfo)
+	if err != nil {
+		wfe.sendError(acme.InternalErrorProblem(fmt.Sprintf("Error marshaling renewalInfo: %s", err)), response)
+		return
+	}
+}
+
+func (wfe *WebFrontEndImpl) determineARIWindow(_ context.Context, id *core.CertID) (*core.RenewalInfo, error) {
+	if id == nil {
+		return nil, errors.New("CertID was nil")
+	}
+
+	// Check if the serial is revoked, and if so, renew it immediately.
+	isRevoked := wfe.db.GetRevokedCertificateBySerial(id.SerialNumber)
+	if isRevoked != nil {
+		return core.RenewalInfoImmediate(time.Now().In(time.UTC)), nil
+	}
+
+	cert := wfe.db.GetCertificateBySerial(id.SerialNumber)
+	if cert == nil {
+		return nil, errors.New("failed to retrieve existing certificate serial")
+	}
+
+	return core.RenewalInfoSimple(cert.Cert.NotBefore, cert.Cert.NotAfter), nil
+}
+
+/*
+func (wfe *WebFrontEndImpl) UpdateRenewal(_ context.Context, response http.ResponseWriter, request *http.Request) {
```

Comment:
This can be removed entirely, no need to attempt to rehydrate it. It doesn't exist in draft-03.